### ISSUE TITLE
Regression with hipSYCL >= 0.8.2 for hand-rolled tree reduction

### ISSUE THREAD

**Issue by tom91136** [AdaptiveCpp/AdaptiveCpp#464]
**Bug summary**
We found a bug with hipSYCL > 0.8.2 where our existing hand-rolled tree reduction in [CloverLeaf](https://github.com/UoB-HPC/cloverleaf_sycl) produced the wrong answer.
This was not the case for versions <= 0.8.2 or DPC++ and ComputeCpp.

**To Reproduce**/**Expected behavior**

To facilitate testing with other SYCL implementations, I've create a [separate repo](https://github.com/UoB-HPC/reduction_sycl).

**Describe your setup**
Programs are compiled using CMake, this bug has only been verified on the CPU.

**Comment by illuhad** [comment]
Thanks for reporting!

I could reproduce the issue, and I think I have found a race condition in your code (see comment in code snippet):

```c++
queue.submit([=](sycl::handler& h) mutable {
    cl::sycl::accessor<double, 1, sycl::access::mode::read_write,
                       sycl::access::target::local>
        local(cl::sycl::range<1>(dotWgsize), h);
    auto drain(result.get_access<sycl::access::mode::read_write>(h));
    h.parallel_for<class reduce>(
        sycl::nd_range<1>(dotNumGroups * dotWgsize, dotWgsize),
        [=](sycl::nd_item<1> item) {
          size_t globalId = item.get_global_id(0);
          size_t localId = item.get_local_id(0);
          size_t globalSize = item.get_global_range()[0];

          local[localId] = DOUBLE_MAX;
          for (; globalId < size; globalId += globalSize) {
            local[localId] = sycl::fmin(local[localId], drain[globalId]);
          }
          size_t localSize = item.get_local_range()[0];
          for (size_t offset = localSize / 2; offset > 0; offset /= 2) {
            item.barrier(sycl::access::fence_space::local_space);
            if (localId < offset) {
              local[localId] =
                  sycl::fmin(local[localId], local[localId + offset]);
            }
          }
          if (localId == 0) {
            // This is where I think the problem is: You are writing to memory at indices [0, dotWgsize).
            // Since we are not guaranteed that groups are synchronized between each other,
            // these data elements may be read simultaneously by your first group trying 
            // to initialize its local memory buffer at the beginning of the kernel.
            drain[item.get_group(0)] = local[0];
          }
        });
  });
```

If I change your code such that partial results for group `i` are stored at  `i*dotWgsize` (which then guarantees that each group only writes to the memory that it processes itself, and no other group reads), the error no longer reproduces for me.

Note: I have only looked at the non-generic version as the error also reproduced there and it was much easier to understand and follow.

**Comment by tom91136** [comment]
Nice catch! 
I wrote that generic reduction back in July 2019 and it's been a source of many bugs for CloverLeaf. 
With SYCL2020 officially released, I'll probably use the newer reduction once all SYCL implementations implement it.

I can confirm your fix does indeed resolved the issue for the non-generic version.
However, while back porting the fix to CloverLeaf, which uses the generic version, I've notice basically the same failure.
Digging further, I was able to recreate the failure in the non-generic version:
```c++
sycl::buffer<double, 1> result = ...
queue.submit([=](sycl::handler& h) {     
    auto source(result.get_access<sycl::access::mode::read>(h));
    auto drain(result.get_access<sycl::access::mode::read_write>(h));
    h.parallel_for<class bar>(sycl::nd_range<1>(42), [=](sycl::nd_item<1> item) {
        // do reduction
        // use drain only, or use both drain and source, doesn't matter
    });
});
// queue.wait();
queue.submit([=](sycl::handler& h) {     
    auto source(result.get_access<sycl::access::mode::read>(h));
    auto drain(result.get_access<sycl::access::mode::read_write>(h));
    h.parallel_for<class bar>(sycl::nd_range<1>(42), [=](sycl::nd_item<1> item) {
        // do group reduction
        // use drain only, or use both drain and source, doesn't matter
    });
});
// queue.wait();
```
The unusual structure (i.e creating accessors but not using them) is an artefact of the generic reduction where I reuse lambdas.
If we then access `result`  at the end, we get the wrong answer most of the time.

The issues only occurs iff:
 * The `source` accessor in the `foo` and `bar` kernel are both set to `read` (any other access mode combination passes verification), whether `source` is actually used in the kernel is not important. 
 * The *absence* of `queue.wait()` or equivalent between the two kernels or at the very end. If we uncomment `queue.wait()` at either location, the issue goes away.

I've updated the reproducer repo with your fix and the condition described above for both the generic and non-generic version.

**Comment by illuhad** [comment]
That was *very* helpful, thank you!

I think I know what the issue is: hipSYCL does various optimizations to the DAG, in particular it elides DAG edges between two operations that only read the same memory, such that the operations may be executed in parallel. I think that in your case, this optimization goes overboard and causes a missing dependency between the two kernels.
What *should* happen is that it detects that there is still a read-write access and therefore still a conflict between the two kernels that requires synchronization. However, we have this innocent looking `if` statement here:

https://github.com/illuhad/hipSYCL/blob/213c683c4fcd3b107d03a6a866c952f7962b0794/src/runtime/dag_builder.cpp#L60

which prevents the update of the users of a memory buffer for conflict detection if there is already an entry for the node. As a consequence, the node remains registered as a read-only access and the DAG edge is removed.

I suspect that probably it might even work if you swap the declaration order of `source` and `drain` in one of the kernels.

What we need to do is remove that if-statement and instead add logic to promote the access mode of existing users if we need stronger guarantees (e.g. read to read-write).

I will verify/come up with a PR tomorrow. Too late already ;)

**Comment by illuhad** [comment]
Fix in PR #466. With this PR, also your updated code no longer reproduces for me.

**Comment by tom91136** [comment]
Thanks for the fix!
Can confirm on my side as well, CloverLeaf also passes verification with the PR.
